### ISSUE TITLE
build: Avoid pushing latest Docker tag for patch releases on old branches

### ISSUE THREAD

**Issue by a-robinson** [cockroachdb/cockroach#25176]
Docker images usually [follow a practice of including a special tag called "latest"](http://container-solutions.com/docker-latest-confusion/) that gets used by default if the user doesn't specify a tag for their image. Our build/push process includes the latest tag for [all stable releases that we push](https://hub.docker.com/r/cockroachdb/cockroach/tags/).

This normally works well, except when we have to push a patch version of an older release, e.g. pushing v1.1.8 after having already released v2.0.0. In such cases, anyone that was depending on getting certain v2.0 behavior when they `docker run cockroachdb/cockroach` could end up with an incompatible v1.1 build.

We need to avoid pushing the `latest` tag if the image we're pushing isn't the latest minor release. I'm not sure there's a good way for us to recognize this from [our current bash script](https://github.com/cockroachdb/cockroach/blob/master/build/teamcity-publish-artifacts.sh), so we may want to convert that to go unless someone (@benesch) has ideas.

Originally reported in https://github.com/cockroachdb/cockroach/issues/22727.

**Comment by a-robinson** [comment]
Related to https://github.com/cockroachdb/cockroach/issues/22756

**Comment by petermattis** [comment]
Let me see what can be done. We've been bitten by this several times.

**Comment by petermattis** [comment]
Here is a shell pipeline cobbled together from ideas on the internets to find the latest version from the tags on docker hub:

```
curl -s -S "https://registry.hub.docker.com/v2/repositories/cockroachdb/cockroach/tags/?Page_size=100" | \
    grep -E -o '"name": "[^"]*"' | grep -v '"latest"' | \
    awk -F\" '{print $4}' | \
    sort -r --version-sort | head -1
```

If we strip of the final component of the version, we can then see if the same major version is being built.

**Comment by petermattis** [comment]
Here is a more fleshed out version of this idea:

```
latestVersion=$(curl -s -S "https://registry.hub.docker.com/v2/repositories/cockroachdb/cockroach/tags/?Page_size=100" | \
        grep -E -o '"name": "[^"]*"' | grep -v '"latest"' | \
        awk -F\" '{print $4}' | \
        sort -r --version-sort | head -1)

function extractMajor() {
    echo $1 | grep -E -o '^v[0-9]+\.[0-9]+\.'
}

for vers in "v19.1.5" "v2.1.9"; do
    if test "$(extractMajor ${latestVersion})" = "$(extractMajor ${vers})"; then
        echo "${vers} is latest"
    else
        echo "${vers} is not latest"
    fi
done
```

We should be able to use `$TC_BUILD_BRANCH` from `teamcity-build-artifacts.sh` and compare against the latest major version tag on docker.

**Comment by rafiss** [comment]
Looks like this was resolved in #41067
